### PR TITLE
Revert stabilization of never type

### LINKED ISSUE #66757: Regression in Error conversion from Infallible
The following code:

```rust
use std::convert::{TryFrom, Infallible};

struct E;

impl From<Infallible> for E {
    fn from(_: Infallible) -> E {
        E
    }
}

fn foo() -> Result<(), E> {
    u32::try_from(1u32)?;
    Ok(())
}
```

[compiles on stable](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=497c2032a83849e5cf80c81e32eccf9c) but [fails to compile on nightly](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=497c2032a83849e5cf80c81e32eccf9c) with:

```
error[E0277]: `?` couldn't convert the error to `E`
  --> src/lib.rs:13:24
   |
13 |     u32::try_from(1u32)?;
   |                        ^ the trait `std::convert::From<()>` is not implemented for `E`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
   = help: the following implementations were found:
             <E as std::convert::From<!>>
   = note: required by `std::convert::From::from`
```

This was discovered through https://github.com/bytecodealliance/wasmtime/issues/633

### PR DESCRIPTION
**Pull request by nikomatsakis**
Fixes https://github.com/rust-lang/rust/issues/66757

I decided to keep the separate `never-type-fallback` feature gate, but tried to otherwise revert https://github.com/rust-lang/rust/pull/65355. Seemed pretty clean.

( cc @Centril, author of #65355, you may want to check this over briefly )

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Actually, one thing that would be nice is to add some regression tests for the pattern in #66757, let me do that.

File: src/librustc_error_codes/error_codes/E0725.md [hunk 1]
```diff
@@ -4,8 +4,8 @@ command line flags.
 Erroneous code example:
 
 ```ignore (can't specify compiler flags from doctests)
-#![feature(specialization)] // error: the feature `specialization` is not in
-                            // the list of allowed features
+#![feature(never_type)] // error: the feature `never_type` is not in
```
Comment (Centril) [inline src/librustc_error_codes/error_codes/E0725.md]: Keep `#![feature(specialization)]` to make things easier the second time around.

**Comment by Centril** [comment]
You'll need to account for https://github.com/rust-lang/rust/pull/65355/files#diff-2fe7e9528ff7e94dc51ea7c62c3c40b9 as well.

r=me when green and ^-- + comment above addressed.

r? @Centril

**Comment by Centril** [comment]
Seems that this broke Clippy in https://github.com/rust-lang/rust/pull/67263#issuecomment-565276447, which is not allowed due to toolstate checks. @bors r-

**Comment by nikomatsakis** [comment]
Regarding clippy, is somebody working on resolving that situation? I saw that you and @Manishearth were discussing earlier.

**Comment by Manishearth** [comment]
Make a PR to clippy that fixes it, we'll add it to a branch, and then you can merge this with a submodule bump.

**Comment by nikomatsakis** [comment]
OK. I should have done this yesterday. =) Oh well. Working on it now. @Manishearth I am asking random questions in #clippy trying to understand how to test this locally. One other question, @Centril, looking at those errors, I think I see some further failures (e.g., to the book etc) -- do you know if the clippy errors are the only ones that will actually *block* the PR from landing?

**Comment by oli-obk** [comment]
I made the clippy changes a branch on the clippy repo (https://github.com/rust-lang/rust-clippy/tree/niko-no-never), so the commit should get picked up correctly by CI

@bors r=centril

**Comment by nikomatsakis** [comment]
@bors p=1

I'm bumping the priority on this because there is time pressure -- we'd like to figure out any complications sooner rather than later. Also, it's Saturday morning and I'd rather have this dealt with.

**Comment by lzutao** [comment]
Is b217587 a work-in-progress?

**Comment by Centril** [comment]
> One other question, @Centril, looking at those errors, I think I see some further failures (e.g., to the book etc) -- do you know if the clippy errors are the only ones that will actually _block_ the PR from landing?

Didn't look that closely sorry; we'll find out soon enough anyways. :)

**Comment by Centril** [comment]
PR builder says it failed so: @bors r- retry

**Pull request merged**